### ISSUE TITLE
dGPU is active when no processes is using it

### ISSUE THREAD

**Issue by WatchMkr** [pop-os/cosmic-comp#590]
Pop!_OS 24.04 | System76 Oryx Pro

![image](https://github.com/pop-os/cosmic-comp/assets/3430283/d76edcde-fc34-4870-9fb3-e5fdb4148064)

![image](https://github.com/pop-os/cosmic-comp/assets/3430283/9e717122-d3a1-4de3-b41b-30c3aaa5f7df)

**Comment by Drakulix** [comment]
I have not yet seen this issue persist (except on a non-system76 with broken firmware, where the gpu never shutdown again, after an external output was once connected until a complete power cycle).

Though I have seen this exact state, when switching ttys. Unfortunately we need quite long delays between reading the current state to not keep the gpu artificially awake, so this only disappeared a good minute later for me.

Does it this state persists longer than a couple of minutes for any of our machines?

**Comment by WatchMkr** [comment]
> Does it this state persists longer than a couple of minutes for any of our machines?

Yes, this the GPU is always active without processes on the Oryx Pro (oryp11).

**Comment by Drakulix** [comment]
> > Does it this state persists longer than a couple of minutes for any of our machines?
> 
> Yes, this the GPU is always active without processes on the Oryx Pro (oryp11).

Hmm, I can't replicate this on my oryp7.

And if nvidia-smi is empty, cosmic-comp shouldn't be to blame here.

Could I get the outputs of the following commands?
- `cat /sys/class/drm/card1/device/power_state` (assuming the nvidia card enumerated to `/dev/dri/card1`)
- `cat /sys/bus/pci/devices/0000\:01\:00.0/power/control` (assuming that is the address on the oryp11 as well)
- `cat /proc/driver/nvidia/gpus/0000\:01\:00.0/power`

**Comment by WatchMkr** [comment]
`cat /sys/class/drm/card1/device/power_state` `D0`
`cat /sys/bus/pci/devices/0000\:01\:00.0/power/control` `on`

`cat /proc/driver/nvidia/gpus/0000\:01\:00.0/power` 
```
Runtime D3 status:          Enabled (fine-grained)
Video Memory:               Active

GPU Hardware Support:
 Video Memory Self Refresh: Supported
 Video Memory Off:          Supported

S0ix Power Management:
 Platform Support:          Not Supported
 Status:                    Disabled
```

**Comment by Drakulix** [comment]
> `cat /sys/bus/pci/devices/0000\:01\:00.0/power/control` `on`

That is the issue.

Something like `echo "auto" | sudo tee /sys/bus/pci/devices/0000\:01\:00.0/power/control` should fix this.

We need to figure out, what causes this value to be "on" instead of "auto".

Nvidia's docs suggest that for Turing and later GPUs this value is "auto" by default (while pre-Turing defaults to "on"): https://download.nvidia.com/XFree86/Linux-x86_64/555.58.02/README/dynamicpowermanagement.html

So this is probably caused by some configuration file installed by some package.

**Comment by WatchMkr** [comment]
`echo "auto" | sudo tee /sys/bus/pci/devices/0000\:01\:00.0/power/control` does indeed fix it until reboot.

**Comment by hgaiser** [comment]
This seems related to https://github.com/pop-os/cosmic-comp/issues/913 . In that case, `cosmic-comp` seems to be running on the dGPU, even though `nvidia-smi` doesn't show any processes.

I don't see that app list in the battery applet to begin with, is it only activated under certain conditions?

![image](https://github.com/user-attachments/assets/ff8f1198-252f-4a25-bd10-889dd626b425)

**Comment by Drakulix** [comment]
> I don't see that app list in the battery applet to begin with, is it only activated under certain conditions?

Only if nvidia-smi reports applications or we are using an output. Otherwise the applet just indicates, that the driver doesn't report the gpu as being shut down. Note that some laptops have buggy firmware, that just keep the GPU enabled indefinitely. There is nothing we can do here, the applet just visualizes this state better, but you will likely see the same issue on gnome. (By looking at the output of `cat "/sys/class/drm/card1/device/power_state"`.)

> This seems related to https://github.com/pop-os/cosmic-comp/issues/913 . In that case, cosmic-comp seems to be running on the dGPU, even though nvidia-smi doesn't show any processes.

See answer at #913, just because cosmic-comp is listed, doesn't mean it is keeping the GPU from powering down. It has logic to free up all related resources.

**Comment by bobcat141** [comment]
> `echo "auto" | sudo tee /sys/bus/pci/devices/0000\:01\:00.0/power/control` does indeed fix it until reboot.

That worked for me, but only after also doing the following:

1- Replace the open kernel NVIDIA drivers with the proprietary ones.
2- Add the following lines to `/etc/modprobe.d/nvidia.conf`:
```
options nvidia "NVreg_EnableGpuFirmware=0"
options nvidia "NVreg_DynamicPowerManagement=0x02"
```
3- Run `sudo update-initramfs -c -k all`

My laptop's dGPU is an NVIDIA 1660TI mobile. I don't know if this helps.
